website/content/docs/rfcs/efficient-search-components.mdx

```diff
@@ -0,0 +1,341 @@
+
+```json
+{ "key": "tag", "value": "adm", "match": "prefix" }
+```
+
+As for the indexes available, they should be provided through a specific endpoint.
+
+## Rationale and Alternatives
+
+### Node Keys
+
+The current implementation uses a simple string-based key for each node, which is sufficient for the initial use case. However, this could be extended to support more complex key structures in the future. Strings keys has the limitation of not ordering integer values.
+
+### Concurrency Control
+
+The current implementation uses tree-level locking with a single `RWMutex` per B+ tree instance. This approach was chosen for its simplicity and correctness guarantees, but it can become a performance bottleneck under high concurrent workloads.
+
+**Alternative Considered: Per-Node Locking**
+
+A more sophisticated approach would implement fine-grained locking at the node level, where each node maintains its own mutex. This would allow multiple concurrent operations to proceed simultaneously as long as they operate on different parts of the tree.
+
+```go
+type Node struct {
+    // ... existing fields
+    lock sync.RWMutex
+}
+```
+
+The tree-level locking approach was selected for the initial implementation to ensure correctness and maintainability. Per-node locking could be considered as a future optimization if benchmarking identifies concurrency as a bottleneck.
+
+### Storage
+
+The current storage system has the limitation of a one-to-one mapping between keys and nodes. This means that each key can only have one node associated with it, which can be a bottleneck when dealing with large amounts of data. To overcome this limitation, a new storage system that supports multiple nodes per key should be implemented.
+
+**Alternative Considered: Page-Based Storage**
+A page-based storage system would allow multiple nodes to be associated with a single key, enabling more efficient storage and retrieval of data. This would require a more complex implementation but would provide better performance for large datasets. The current implementation uses a simple key-value store, which is sufficient for the initial use case but may need to be replaced in the future as the system scales.
+
+## Downsides
+
+Code complexity and additional load on the system and storage when building indexes. It is suggested to make index building opt-in and also tunable so that system administrators can disable index building for mounts if desired.
+
+## Security Implications
+
+No security implications were identified.
+
+## User/Developer Experience
+
+With these constructs available, an OpenBao developer will be able to dynamically configure and build indexes on the data persisted in storage.
+No user facing changes are expected with this RFC as it only introduces the constructs to build the indexes.
+
+## Unresolved Questions
+
+- How to keep track of the current progress of an index that is being incrementally built;
```

Comment:
I think this would have to be in the plugin. They'd add an `indexed: <bool>` on the items themselves or an `indexed: <int>` for versioned items (with async re-indexing). Linear scan is going to be the only approach, +/- a queue system for handling async updates.

Reply:
Not sure if  I understood what you are proposing here. The idea is for the plugin to track the items indexed but wouldn't that be iterating over all items and check if they have already been indexed (by checking the label, tag)? I was thinking more along the lines of having a milestone but don't know how would that be feasible.